### PR TITLE
Bump SwarmKit to f74983e7c015a38a81c8642803a78b8322cf7eac

### LINKED ISSUE #33795: swarm mode duplicate ip addresses
**Description**
We recently had several issues with our cloud provider which lead us to several docker swarms failing with strange errors. One thing we observied that from time to time IP adresses in the swarm overlay network are assigned duplicate. 
We made several attempts to reproduce this issue and finally found a way to reproduce the duplicate ip adress issue. 

**Steps to reproduce the issue:**
1. Create a docker swarm with several nodes (five in our example)
2. Start some Services in the swarm with at least two replicas
3. Start some containers on swarm nodes, that are attached to the swarm overlay network
4. kill the docker process (kill -9 <main_pid_from systemctl status docker>) and/or kill some service tasks running on the same node (kill -9 <main_pid_from docker inspect>)
5. do a "docker network inspect -v <overlay_network>" on the node
6. start and stop some services in the swarm

**Describe the results you received:**
I can see that ip adresses are assigned to more than one service task. 

**Describe the results you expected:**
I would expect that every service task is assigned a unique ip adress.

**Additional information you deem important (e.g. issue happens only occasionally):**

What we tried to resolve this issue: 
- let affected node leave the swarm (docker swarm leave, docker node demote, docker node rm)
- stop the docker daemon on the affected node
- rm -rf  /var/lib/docker/*
- start the docker daemon
- rejoin the swarm

there are still duplicate ip adresses (ocker network inspect -v <overlay_network>)

next try: 
- let node leave the swarm again 
- stop the host
- delete the host
- create a new host (new host name, new ip address)
- let new host join swarm as manager

again, there are duplicate ip adresses which are assigned to the host that was completely removed from the swarm (see attached file overlay.txt)
[overlay.txt](https://github.com/moby/moby/files/1097600/overlay.txt)
**Output of `docker version`:**
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64
 Experimental: false

```
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

**Output of `docker info`:**
see attached file docker_info.txt
[docker_info.txt](https://github.com/moby/moby/files/1097641/docker_info.txt)
**Additional environment details (AWS, VirtualBox, physical, etc.):**

Environment: 
OS: CentOS 7 running in OpenStack in a public cloud.

### PR DESCRIPTION
**Pull request by thaJeztah**
Full diff:

https://github.com/docker/swarmkit/compare/68a376dc30d8c4001767c39456b990dbd821371b...f74983e7c015a38a81c8642803a78b8322cf7eac

- Replace EC Private Key with PKCS#8 PEMs https://github.com/docker/swarmkit/pull/2246
- Fix IP overlap with empty EndpointSpec https://github.com/docker/swarmkit/pull/2505 (fixes https://github.com/moby/moby/issues/33795)
- Add support for Support SCTP port mapping (depends on changes in libnetwork) https://github.com/docker/swarmkit/pull/2298
- [orchestrator/updater] Do not reschedule tasks if only placement constraints change and are satisfied by the assigned node https://github.com/docker/swarmkit/pull/2496
- Ensure task reaper stopChan is closed no more than once https://github.com/docker/swarmkit/pull/2491
- [manager/dispatcher] Synchronization fixes (https://github.com/docker/swarmkit/pull/2495)

ping @anshulpundir @nishanttotla PTAL

### CODE REVIEW

**Review by aaronlehmann — APPROVED** [review]

**Comment by thaJeztah** [comment]
looks like several tests are failing

**Comment by thaJeztah** [comment]
Failures are due to the changes in https://github.com/docker/swarmkit/pull/2246, so the tests may need some updating. Talked with @cyli on Slack, and she'll assist with updating the tests 👍

**Comment by cyli** [comment]
@thaJeztah Tested this locally, and the swarm lock/unlock tests that failed all pass with this change:

```diff
diff --git a/integration-cli/docker_cli_swarm_test.go b/integration-cli/docker_cli_swarm_test.go
index 283f576f5..4a5ec9a56 100644
--- a/integration-cli/docker_cli_swarm_test.go
+++ b/integration-cli/docker_cli_swarm_test.go
@@ -4,7 +4,6 @@ package main

 import (
        "bytes"
-       "crypto/x509"
        "encoding/json"
        "encoding/pem"
        "fmt"
@@ -25,6 +24,7 @@ import (
        "github.com/docker/libnetwork/driverapi"
        "github.com/docker/libnetwork/ipamapi"
        remoteipam "github.com/docker/libnetwork/ipams/remote/api"
+       "github.com/docker/swarmkit/ca/keyutils"
        "github.com/go-check/check"
        "github.com/gotestyourself/gotestyourself/fs"
        "github.com/gotestyourself/gotestyourself/icmd"
@@ -1007,7 +1007,7 @@ func checkKeyIsEncrypted(d *daemon.Swarm) func(*check.C) (interface{}, check.Com
                        return fmt.Errorf("invalid PEM-encoded private key"), nil
                }

-               return x509.IsEncryptedPEMBlock(keyBlock), nil
+               return keyutils.IsEncryptedPEMBlock(keyBlock), nil
        }
 }
```

**Comment by thaJeztah** [comment]
Applied the patch; let's see if CI goes green now 👍

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

File: vendor/github.com/docker/swarmkit/manager/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -532,6 +539,21 @@ func (d *Dispatcher) register(ctx context.Context, nodeID string, description *a
...
 // on every status change of its tasks.
 func (d *Dispatcher) UpdateTaskStatus(ctx context.Context, r *api.UpdateTaskStatusRequest) (*api.UpdateTaskStatusResponse, error) {
+	// shutdownWait.Add() followed by isRunning() to ensures that
+	// if this rpc sees the dispatcher running,
+	// it will already have called Add() on the shutdownWait wait,
+	// which ensures that Stop() will wait for this rpc to complete.
+	// Note that Stop() first does Dispatcher.ctx.cancel() followed by
+	// shutdownWait.Wait() to make sure new rpc's don't start before waiting
+	// for existing ones to finish.
+	d.shutdownWait.Add(1)
```
Comment (cpuguy83) [inline vendor/github.com/docker/swarmkit/manager/dispatcher/dispatcher.go]: This looks suspect to me.
A waitgroup `Add()` is not safe to do async.
Ping @anshulpundir

**Review by cpuguy83 — COMMENTED** [review]
Looks like this patch is pulling in a change from https://github.com/docker/swarmkit/pull/2495 that I'm a bit suspect of. The usage of the waitgroup looks racey on first glance.

File: vendor/github.com/docker/swarmkit/manager/dispatcher/dispatcher.go (hunk 1 above)
Comment (anshulpundir) [reply vendor/github.com/docker/swarmkit/manager/dispatcher/dispatcher.go]: sync.WaitGroup operations should be atomic. I think this is a false alarm. I will try to follow up with the community.

FYI: https://groups.google.com/forum/#!topic/golang-nuts/Qq_h0_M51YM

Comment (anshulpundir) [reply vendor/github.com/docker/swarmkit/manager/dispatcher/dispatcher.go]: Actually it seems Add() needs to be synchronized with Wait(). Bummer.

I'll look at how to fix this.

**Review by anshulpundir — APPROVED** [review]
@nishanttotla and I discussed the data race failure in unit-tests and it should be OK to move forward. Because of how the waitgroup is used, the race should not have any undesired side-effects.

**Review by vdemeester — APPROVED** [review]
LGTM 🌵

**Comment by thaJeztah** [comment]
Looks like https://github.com/docker/swarmkit/pull/2495 is being reverted in https://github.com/docker/swarmkit/pull/2518

**Pull request merged**